drivers/pinctrl/pinctrl_mchp_xec.c

```diff
@@ -0,0 +1,172 @@
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_240_276)),
+};
+
+/** Number of GPIO ports. */
+static const size_t gpio_ports_cnt = ARRAY_SIZE(gpio_ports);
+
+/* Microchip XEC: each GPIO pin has two 32-bit control register.
+ * The first 32-bit register contains all pin features except
+ * slew rate and driver strength in the second control register.
+ * We compute the register index from the beginning of the GPIO
+ * control address space which is the same range of the PINCTRL
+ * parent node.
+ */
+
+static void config_drive_slew(uint32_t port, gpio_pin_t pin, uint32_t conf)
+{
+	struct gpio_regs * const regs = (struct gpio_regs * const)DT_INST_REG_ADDR(0);
+	uint32_t idx = (port * 32U) + pin;
+	uint32_t slew = conf &
+		(MCHP_XEC_OSPEEDR_MASK << MCHP_XEC_OSPEEDR_POS);
+	uint32_t drvstr = conf &
+		(MCHP_XEC_ODRVSTR_MASK << MCHP_XEC_ODRVSTR_POS);
+	uint32_t val = 0;
+	uint32_t mask = 0;
+
+	if (slew != MCHP_XEC_OSPEEDR_NO_CHG) {
+		mask |= MCHP_GPIO_CTRL2_SLEW_MASK;
+		if (slew == MCHP_XEC_OSPEEDR_FAST) {
+			val |= MCHP_GPIO_CTRL2_SLEW_FAST;
+		}
+	}
+	if (drvstr != MCHP_XEC_ODRVSTR_NO_CHG) {
+		mask |= MCHP_GPIO_CTRL2_DRV_STR_MASK;
+		val |= (drvstr << MCHP_GPIO_CTRL2_DRV_STR_POS);
+	}
+
+	if (!mask) {
+		return;
+	}
+
+	regs->CTRL2[idx] = (regs->CTRL2[idx] & ~mask) | (val & mask);
+}
+
+static void config_alt_func(uint32_t port, gpio_pin_t pin, uint32_t altf)
+{
+	struct gpio_regs * const regs = (struct gpio_regs * const)DT_INST_REG_ADDR(0);
+	uint32_t idx = (port * 32U) + pin;
+	uint32_t muxval = (uint32_t)((altf & MCHP_GPIO_CTRL_MUX_MASK0) <<
+				     MCHP_GPIO_CTRL_MUX_POS);
+
+	regs->CTRL[idx] = (regs->CTRL[idx] & ~(MCHP_GPIO_CTRL_MUX_MASK)) | muxval;
+}
+
+/* Translate PINCTRL pin configuration to GPIO driver pin configuration
+ * and call GPIO driver to configure the pin.
+ */
+static int xec_config_pin(uint32_t portpin, uint32_t conf, uint32_t altf)
+{
+	uint32_t port = MCHP_XEC_PINMUX_PORT(portpin);
+	gpio_pin_t pin = (gpio_pin_t)MCHP_XEC_PINMUX_PIN(portpin);
+	gpio_flags_t flags = 0;
+	uint32_t temp = 0;
+	int ret = 0;
+
+	if (port >= gpio_ports_cnt) {
+		return -EINVAL;
+	}
+
+	temp = (conf & MCHP_XEC_PUPDR_MASK) >> MCHP_XEC_PUPDR_POS;
+	switch (temp) {
+	case MCHP_XEC_PULL_UP:
+		flags |= GPIO_PULL_UP;
+		break;
+	case MCHP_XEC_PULL_DOWN:
+		flags |= GPIO_PULL_DOWN;
+		break;
+	case MCHP_XEC_REPEATER:
+		flags |= (GPIO_PULL_UP | GPIO_PULL_DOWN);
+		break;
+	default:
+		break;
+	}
+
+	if ((conf >> MCHP_XEC_OTYPER_POS) & MCHP_XEC_OTYPER_MASK) {
+		flags |= GPIO_LINE_OPEN_DRAIN;
+	}
+
+	temp = (conf >> MCHP_XEC_OVAL_POS) & MCHP_XEC_OVAL_MASK;
+	if (temp) {
+		flags |= GPIO_OUTPUT;
+		if (temp == MCHP_XEC_OVAL_DRV_HIGH) {
+			flags |= GPIO_OUTPUT_INIT_HIGH;
+		} else {
+			flags |= GPIO_OUTPUT_INIT_LOW;
+		}
+	} else {
+		flags |= GPIO_INPUT;
+	}
+
+	if (conf & BIT(MCHP_XEC_PIN_LOW_POWER_POS)) {
+		flags = GPIO_DISCONNECTED;
+	}
+
+	ret = gpio_pin_configure(gpio_ports[port], pin, flags);
+	if (ret) {
+		return ret;
+	}
```

Comment:
please don't mix pinctrl/GPIO. pinctrl driver should be self-contained, so the logic to configure should be in this driver.

Reply:
we are making a API call to gpio driver; the function gpio_xec_configure is quite big, don't want to duplicate the code in pinctrl.

Reply:
it's 87 LoC, whereas pinctrl "wrapper" is 61. I'd qualify as "quite big", so please make pinctrl driver independent.

Reply:
@gmarull 
Could we take up pinctrl driver independent of gpio driver as a new PR? 
We would like to get all other pinctrl changes approved and merged first.

Reply:
I see no reasons to keep a GPIO based pinctrl driver, so it's a -1 from my side. I already explained the reasons in the previous PR.

Reply:
updated pinctrl to be independent of gpio driver; please review and approve.